flink/src/main/java/org/apache/iceberg/flink/data/vectorized/VectorizedFlinkOrcReaders.java

```diff
@@ -0,0 +1,278 @@
+      }
+
+      return new FlinkRowColumnVector(fieldVectors, structVector);
+    }
+
+    private ColumnVector getConstantVectors(Type type, Object constant) {
+      if (type.isPrimitiveType()) {
+        Type.TypeID typeID = type.typeId();
+        switch (typeID) {
+          case INTEGER:
+          case DATE:
+          case TIME:
+            return ConstantColumnVectors.ints(constant);
+
+          case LONG:
+            return ConstantColumnVectors.longs(constant);
+
+          case BOOLEAN:
+            return ConstantColumnVectors.booleans(constant);
+
+          case DOUBLE:
+            return ConstantColumnVectors.doubles(constant);
+
+          case FLOAT:
+            return ConstantColumnVectors.floats(constant);
+
+          case DECIMAL:
+            return ConstantColumnVectors.decimals(constant);
+
+          case TIMESTAMP:
+            return ConstantColumnVectors.timestamps(constant);
+
+          case STRING:
+            return ConstantColumnVectors.bytes(constant);
+
+          default:
+            throw new UnsupportedOperationException("Unsupported data type for constant.");
+        }
+      } else {
+        throw new UnsupportedOperationException("ConstantColumnVector only supports primitives.");
+      }
+    }
+
+    private static class ArrayConverter implements Converter {
+      private final Converter elementConverter;
+
+      private ArrayConverter(Converter elementConverter) {
+        this.elementConverter = elementConverter;
+      }
+
+      @Override
+      public ColumnVector convert(org.apache.orc.storage.ql.exec.vector.ColumnVector vector, int batchSize,
+                                  long batchOffsetInFile) {
+        ListColumnVector listVector = (ListColumnVector) vector;
+        ColumnVector elementVector = elementConverter.convert(listVector.child, batchSize, batchOffsetInFile);
+
+        return new ArrayColumnVector() {
+          @Override
+          public ArrayData getArray(int rowId) {
+            int index = getRowIndex(rowId);
+            return new ColumnarArrayData(elementVector, (int) listVector.offsets[index],
+                (int) listVector.lengths[index]);
+          }
+
+          @Override
+          public boolean isNullAt(int rowId) {
+            return vector.isNull[getRowIndex(rowId)];
+          }
+
+          private int getRowIndex(int rowId) {
+            return vector.isRepeating ? 0 : rowId;
+          }
+        };
+      }
+    }
+  }
+
+  private static class FlinkRowColumnVector implements RowColumnVector {
+
+    private ColumnVector[] fieldVectors;
+    private StructColumnVector structVector;
+
+    FlinkRowColumnVector(ColumnVector[] fieldVectors,
+                         StructColumnVector structVector) {
+      this.fieldVectors = fieldVectors;
+      this.structVector = structVector;
+    }
+
+    @Override
+    public ColumnarRowData getRow(int i) {
+      VectorizedColumnBatch vectorizedColumnBatch = new VectorizedColumnBatch(fieldVectors);
```

Comment:
I don't think we need to create a separate `VectorizedColumnBatch` for every row, that's too wasting. How about moving this line into the `FlinkRowColumnVector` constructor ?

Reply:
I updated it